Fix logic in cucushift-installer-check-reboot-nodes-commands.sh for OCP-23070 - [upi-on-…

liweinan opened the pull request:
…aws] Install, reboot and destroy a cluster

liweinan commented:
/pj-rehearse pull-ci-openshift-aws-ebs-csi-driver-release-4.20-e2e-aws-csi-extended

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
# OpenShift Cluster Node Reboot Test Analysis Report

- https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/69255/rehearse-69255-periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28/1968551649320374272
- https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/69255/rehearse-69255-periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28/1968551649320374272/artifacts/aws-upi-reboot-nodes-f28/cucushift-installer-check-reboot-nodes/build-log.txt

## Executive Summary

**✅ All 6 cluster nodes successfully rebooted and cluster fully recovered**

This report analyzes the build log from the `cucushift-installer-check-reboot-nodes` test execution, demonstrating successful node reboot operations and cluster recovery.

## Test Execution Overview

- **Test Duration**: 1,015 seconds (approximately 17 minutes)
- **Total Nodes**: 6 nodes (3 workers + 3 masters)
- **Success Rate**: 100% (all nodes successfully rebooted)
- **Cluster Recovery**: Complete and healthy

## Node Reboot Statistics

### Successfully Rebooted Nodes

**✅ All 6 nodes completed successful reboot cycles:**

#### Worker Nodes (3 nodes)
1. `ip-10-0-49-123.ec2.internal` (Instance: i-0f6075c6638aff90b) ✅
2. `ip-10-0-64-242.ec2.internal` (Instance: i-066dd76cafa1ed4c8) ✅  
3. `ip-10-0-83-127.ec2.internal` (Instance: i-03ee09c9648d21388) ✅

#### Master Nodes (3 nodes)
1. `ip-10-0-56-103.ec2.internal` (Instance: i-0b2549b8a6692d1b0) ✅
2. `ip-10-0-79-124.ec2.internal` (Instance: i-0c5e6133ed3434ac7) ✅
3. `ip-10-0-83-237.ec2.internal` (Instance: i-0eb7b2410ef9d7259) ✅

## Reboot Process Timeline

| Node | Start Time | Stop Time | Start Time | Total Duration |
|------|------------|-----------|------------|----------------|
| Worker 1 | 06:45:44 | 06:47:52 | 06:48:08 | ~2m 24s |
| Worker 2 | 06:48:08 | 06:49:40 | 06:49:56 | ~1m 48s |
| Worker 3 | 06:49:56 | 06:52:25 | 06:52:41 | ~2m 45s |
| Master 1 | 06:52:41 | 06:54:39 | 06:54:56 | ~2m 15s |
| Master 2 | 06:54:56 | 06:57:36 | 06:57:52 | ~2m 56s |
| Master 3 | 06:57:52 | 07:00:09 | 07:00:25 | ~2m 33s |

## Cluster Recovery Verification

### API Server Accessibility
- **Status**: ✅ Immediately available
- **Time**: 07:00:25 (right after last node restart)

### Node Readiness
- **Initial State**: 5 Ready, 1 NotReady (ip-10-0-83-237.ec2.internal)
- **Recovery Time**: 1 minute 13 seconds (07:00:25 → 07:01:38)
- **Final State**: All 6 nodes Ready

### Cluster Operators Stability
- **Recovery Time**: 1 minute (07:01:38 → 07:02:38)
- **Final State**: All 30 operators stable (True, False, False pattern)

## Key Observations

### Reboot Sequence
- **Pattern**: Worker nodes first, then master nodes
- **Strategy**: Sequential reboot to maintain cluster stability

### Recovery Performance
- **Speed**: Fast cluster recovery (~2 minutes total)
- **Reliability**: No failed operations or timeouts
- **Monitoring**: Detailed status tracking throughout process

### Enhanced Logging
The refactored script provided comprehensive status visibility:
- Real-time instance state transitions
- Periodic status updates during waiting periods
- Success confirmation with detailed status output
- Final cluster health verification

## Final Cluster Status

### Node Status
```
NAME                          STATUS   ROLES                  AGE   VERSION
ip-10-0-49-123.ec2.internal   Ready    worker                 40m   v1.33.4
ip-10-0-56-103.ec2.internal   Ready    control-plane,master   55m   v1.33.4
ip-10-0-64-242.ec2.internal   Ready    worker                 40m   v1.33.4
ip-10-0-79-124.ec2.internal   Ready    control-plane,master   56m   v1.33.4
ip-10-0-83-127.ec2.internal   Ready    worker                 40m   v1.33.4
ip-10-0-83-237.ec2.internal   Ready    control-plane,master   56m   v1.33.4
```

### Cluster Operators
- **Total Operators**: 30
- **Status**: All stable (Available=True, Progressing=False, Degraded=False)
- **Key Operators**: authentication, etcd, kube-apiserver, network, storage

### Machine Config Pools
- **Master Pool**: 3/3 machines ready
- **Worker Pool**: 3/3 machines ready
- **Status**: All pools updated and ready

## Performance Metrics

| Metric | Value |
|--------|-------|
| Total Execution Time | 1,015 seconds (17 minutes) |
| Node Reboot Success Rate | 100% (6/6) |
| Cluster Recovery Time | ~2 minutes |
| API Server Downtime | < 1 minute |
| Final Exit Code | 0 (Success) |

## Conclusion

**✅ Test Result: PASSED**

The node reboot test completed successfully with the following achievements:

1. **Complete Node Coverage**: All 6 cluster nodes were successfully rebooted
2. **Fast Recovery**: Cluster recovered within 2 minutes of the last node restart
3. **Health Verification**: All components (nodes, operators, pools) returned to healthy state
4. **Enhanced Monitoring**: The refactored script provided excellent visibility into the reboot process
5. **Reliable Operations**: No failures, timeouts, or error conditions encountered

The test demonstrates that the OpenShift cluster can handle node reboots gracefully and recover quickly, validating the robustness of the cluster infrastructure and the effectiveness of the reboot testing procedures.

## Recommendations

1. **Continue Monitoring**: The enhanced logging capabilities should be maintained for future tests
2. **Performance Baseline**: This 17-minute total execution time can serve as a baseline for future comparisons
3. **Documentation**: The detailed status tracking should be documented for troubleshooting purposes
4. **Automation**: Consider automating this test as part of regular cluster health validation

---

*Report generated on: 2025-09-18*  
*Test Environment: OpenShift 4.20.0-0.nightly-2025-09-15-100049*  
*Cluster Type: AWS UPI (User-Provisioned Infrastructure)*

liweinan commented:
# # Reboot Script Enhancement Report

## Summary
Enhanced the OpenShift cluster reboot test script with improved reliability, error handling, and monitoring capabilities while maintaining the original reboot logic.

## Key Changes

### 1. Reboot Method Enhancement
- **Before**: SSH-based `sudo reboot` commands
- **After**: AWS EC2 API stop/start operations
- **Benefit**: Eliminates SSH connectivity issues and provides more reliable node restart

### 2. Error Handling Improvements
- Added comprehensive error handling with detailed logging
- Disabled `set -o errexit` to prevent premature script termination
- Implemented ERR and EXIT traps for better error tracking
- Added cluster diagnostics on failure

### 3. Enhanced Monitoring & Logging
- Standardized timestamp format across all log messages
- Added execution time tracking and process information
- Implemented `wait_for_condition` function for robust waiting logic
- Added detailed status reporting during wait periods

### 4. Improved Wait Logic
- **API Server**: 15-minute timeout with 30-second intervals
- **Node Ready**: 90-minute timeout with 60-second intervals  
- **Cluster Operators**: 120-minute timeout with 60-second intervals
- More precise node status checking using JSON parsing

### 5. Configuration Updates
- Increased script timeout to 3 hours
- Added 5-minute grace period
- Enhanced AWS credential and region handling

## Core Logic Preservation
✅ Node discovery (master/worker identification)  
✅ Compact cluster support  
✅ Sequential node restart process  
✅ Cluster recovery waiting  
✅ Final status verification  

## Benefits
- **Reliability**: AWS API-based restart eliminates SSH failures
- **Observability**: Comprehensive logging and diagnostics
- **Robustness**: Better error handling and recovery
- **Maintainability**: Cleaner code structure with reusable functions

The enhancement maintains 100% backward compatibility while significantly improving script reliability and debuggability.

liweinan commented:
@pj-rehearse pull-ci-openshift-aws-ebs-csi-driver-master-e2e-aws-csi-extended

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-ibmcloud-ipi-private-sshkey-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-azure-upi-proxy-reboot-nodes-f14

liweinan commented:
Auzure platform reboot checking looks good: https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/69255/rehearse-69255-periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-azure-upi-proxy-reboot-nodes-f14/1971138058069741568/artifacts/azure-upi-proxy-reboot-nodes-f14/cucushift-installer-check-reboot-nodes/build-log.txt

liweinan commented:
The test failure is not relevant to the changes in this PR.

gpei commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
```diff
@@ -52,32 +353,106 @@ function reboot_cluster() {
+        
+        # Wait fixed time for all nodes to stop
+        echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") - Waiting 2 minutes for all nodes to stop completely..."
+        sleep 120
```
can we use `aws` command to check/wait for all nodes to be stopped, rather than using a fixed sleep time?

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
INFRA ID is the unique identifier for each cluster, so please use that filter the instances

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
We have a health check script, can we use that instead?

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
I think `SOFT_REBOOT` and `HARD_REBOOT` are better:
* For AWS platform (in this PR), the default is `HARD_REBOOT`, operate instances via the cloud command directly.
* for other platforms, the default is `SOFT_REBOOT`, keep existing behaviors, we may create a follow-up PR to support `HARD_REBOOT` on these platform later.

yunjiang29 reviewed (commented):
@liweinan

yunjiang29 commented:
cc @jinyunma @jianli-wei

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-ibmcloud-ipi-private-sshkey-f28 periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-azure-mag-ipi-fullyprivate-reboot-nodes-f14

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-azure-mag-ipi-fullyprivate-reboot-nodes-f14

liweinan commented:
@yunjiang29 I have refactored the commit according to your review. I checked the build log and confirmed that the tests related to these changes passed successfully. The subsequent test failures are unrelated. Could you please review it again when you have time? Thanks!

liweinan commented:
I have to add a `wait_for_cluster_recovery` function to do a simple health check, or the following steps failed because the cluster has not been fully recovered:

- https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/69255/rehearse-69255-periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28/1982099963198312448 

```
INFO[2025-10-25T16:03:47Z] Logs for container test in pod aws-upi-reboot-nodes-f28-enable-qe-catalogsource: 
INFO[2025-10-25T16:03:47Z] no proxy setting.
Running Command: oc whoami
The connection to the server api.ci-op-sx5yyqpm-6b24f.qe.devcluster.openshift.com:6443 was refused - did you specify the right host or port?
{"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"sigs.k8s.io/prow/pkg/entrypoint/run.go:84","func":"sigs.k8s.io/prow/pkg/entrypoint.Options.internalRun","level":"error","msg":"Error executing test process","severity":"error","time":"2025-10-25T16:03:46Z"}
error: failed to execute wrapped command: exit status 1 
INFO[2025-10-25T16:03:47Z] Step aws-upi-reboot-nodes-f28-enable-qe-catalogsource failed after 8s. 
INFO[2025-10-25T16:03:47Z] Step phase pre failed after 1h1m5s.          
```

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-azure-mag-ipi-fullyprivate-reboot-nodes-f14

yunjiang29 commented on ci-operator/config/openshift/verification-tests/openshift-verification-tests-main__installation-nightly-4.20.yaml:
```diff
@@ -454,6 +454,7 @@ tests:
     env:
       BASE_DOMAIN: qe.devcluster.openshift.com
       ENABLE_REBOOT_CHECK: "true"
+      REBOOT_TYPE: HARD_REBOOT
```
I remember we have aws-ipi-reboot job, we need to config HARD_REBOOT for it as well.

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
```diff
@@ -4,10 +4,376 @@ set -o nounset
+        
+        # Determine AWS credentials file location
+        aws_cred_file=""
+        if [[ -n "${CLUSTER_PROFILE_DIR:-}" && -f "${CLUSTER_PROFILE_DIR}/.awscred" ]]; then
```
`${CLUSTER_PROFILE_DIR}/.awscred` is the only location where the AWS credential file is stored, so we don't need to check it.

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
```diff
@@ -4,10 +4,376 @@ set -o nounset
+        # Extract AWS credentials from the credential file (for CI environment)
+        if [[ "${aws_cred_file}" == *"/.awscred" ]]; then
+            AWS_ACCESS_KEY_ID=$(cat "${aws_cred_file}" | grep aws_access_key_id | tr -d ' ' | cut -d '=' -f 2)
+            AWS_SECRET_ACCESS_KEY=$(cat "${aws_cred_file}" | grep aws_secret_access_key | tr -d ' ' | cut -d '=' -f 2)
```
The above `AWS_SHARED_CREDENTIALS_FILE` is satisfied to communicate with AWS services, we usually don't use `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
```diff
@@ -4,10 +4,376 @@ set -o nounset
+    if ! wait_for_condition "API server connectivity" \
+        "oc whoami >/dev/null 2>&1" \
+        "${CLUSTER_RECOVERY_TIMEOUT}" 30 \
+        "API server not accessible within 60 minutes" \
```
If the API is not accessible, the next oc command or the next health check step would fail. So an individual API-check step is not necessary

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
```diff
@@ -4,10 +4,376 @@ set -o nounset
+    # Step 2: Wait for all nodes to be Ready
+    log "Step 2: Waiting for all nodes to be Ready..."
+    if ! wait_for_condition "all nodes to be Ready" \
+        "oc wait nodes --all --for=condition=Ready=true --timeout=30s >/dev/null 2>&1" \
```
`oc wait nodes` command is enough for the node checking, right? Do we need to pass the `wait` command to another `wait` (wait_for_condition) command?

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-upi-reboot-nodes-f28 periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-aws-ipi-proxy-reboot-nodes-arm-f28

yunjiang29 commented on ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh:
```diff
@@ -4,10 +4,305 @@ set -o nounset
+    log "Using AWS hard reboot - simplified logic"
+    
+    # Get all instances for the cluster
+    if ! all_instances=$(get_all_aws_instance_ids); then
```
```suggestion
    all_instances=$(aws ec2 describe-instances --region $AWS_REGION --filters "Name=tag-key,Values=kubernetes.io/cluster/${infra_id}" 'Name=instance-state-name,Values=running,stopped,pending,stopping' --query 'Reservations[*].Instances[*].InstanceId' --output text | tr '\n' ' ' | sed 's/ $//')
    if [ "$all_instances" == "" ]; then
```

yunjiang29 commented:
@liweinan you can rehearse another soft reboot cluster to verify your PR.

liweinan commented:
/pj-rehearse periodic-ci-openshift-verification-tests-main-installation-nightly-4.20-azure-mag-ipi-fullyprivate-reboot-nodes-f14

liweinan commented:
@yunjiang29 I have added the `HARD_REBOOT` option into `4.21` ipi + upi tests and checked the build log and all pasted. So I squashed the commits. Could you please help check whether this can be merged? Thanks!

yunjiang29 commented:
/pj-rehearse ack

yunjiang29 commented:
/lgtm

yunjiang29 commented:
@gpei PTAL
